src/Std/Time/DateTime/Timestamp.lean

```diff
@@ -0,0 +1,224 @@
+  t.second
+
+/--
+Negates a `Timestamp`, flipping its second and nanosecond values.
+-/
+@[inline]
+def neg (t : Timestamp) : Timestamp := by
+  refine ⟨-t.second, t.nano.neg, ?_⟩
+  cases t.proof with
+  | inl n => exact Or.inr (n.imp Int.neg_le_neg Int.neg_le_neg)
+  | inr n => exact Or.inl (n.imp Int.neg_le_neg Int.neg_le_neg)
+
+/--
+Adds two timestamps together, handling any carry-over in nanoseconds.
+-/
+def add (t₁ t₂ : Timestamp) : Timestamp := by
+  let diffSecs := t₁.second + t₂.second
+  let diffNano := t₁.nano.addBounds t₂.nano
+
+  let (diffSecs, diffNano) : Second.Offset × Nanosecond.Span :=
+    if h₀ : diffNano.val > 999999999 then
+      have diffNano := diffNano.truncateBottom h₀ |>.sub 999999999
+      (diffSecs + 1, diffNano.expandBottom (by decide))
+    else if h₁ : diffNano.val < -999999999 then
+      have diffNano := diffNano.truncateTop (Int.le_sub_one_of_lt h₁) |>.add 999999999
+      (diffSecs - 1, diffNano.expandTop (by decide))
+    else by
+      have h₀ := Int.le_sub_one_of_lt <| Int.not_le.mp h₀
+      have h₁ := Int.le_sub_one_of_lt <| Int.not_le.mp h₁
+      simp_all [Int.add_sub_cancel]
+      exact ⟨diffSecs, Bounded.mk diffNano.val (And.intro h₁ h₀)⟩
+
+  if h : diffSecs.val > 0 ∧ diffNano.val < 0 then
+    let truncated := diffNano.truncateTop (Int.le_sub_one_of_lt h.right)
+    let nano := truncated.addTop 1000000000 (by decide)
+    let proof₁ : 0 ≤ diffSecs - 1 := Int.le_sub_one_of_lt h.left
+    refine { second := UnitVal.mk (diffSecs.val - 1), nano, proof := ?_ }
+    simp [nano, Bounded.LE.addTop]
+    refine (Or.inl (And.intro proof₁ ?_))
+    let h₃ := (Int.add_le_add_iff_left 1000000000).mpr diffNano.property.left
+    rw [Int.add_comm]
+    exact Int.le_trans (by decide) h₃
+  else if h₁ : diffSecs.val < 0 ∧ diffNano.val > 0 then
+    let second := diffSecs.val + 1
+    let truncated := diffNano.truncateBottom h₁.right
+    let nano := truncated.subBottom 1000000000 (by decide)
+    refine { second := UnitVal.mk second, nano, proof := ?_ }
+    simp [nano, truncated, Bounded.LE.subBottom, Bounded.LE.truncateBottom]
+    refine (Or.inr (And.intro ?_ ?_))
+    · exact h₁.left
+    · let h₃ := Int.sub_le_sub_right diffNano.property.right 1000000000
+      simp at h₃
+      exact Int.le_trans h₃ (by decide)
+  else
+    refine ⟨diffSecs, diffNano, ?_⟩
+    if h₂ : diffSecs.val > 0 then
+      exact Or.inl (And.intro (Int.le_of_lt h₂) (Int.not_lt.mp (not_and.mp h h₂)))
+    else if h₃ : diffSecs.val < 0 then
+      exact Or.inr (And.intro (Int.le_of_lt h₃) (Int.not_lt.mp (not_and.mp h₁ h₃)))
+    else
+      exact Int.le_total diffNano.val 0 |>.symm.imp (And.intro (Int.not_lt.mp h₃)) (And.intro (Int.not_lt.mp h₂))
+
+/--
+Subtracts one `Timestamp` from another.
+-/
+@[inline]
+def sub (t₁ t₂ : Timestamp) : Timestamp :=
+  t₁.add t₂.neg
+
+/--
+Creates a new `Timestamp` out of `Second.Offset`.
+-/
+@[inline]
+def ofSeconds (s : Second.Offset) : Timestamp := by
```

Comment:
I think that this should be called something like `ofSecondsSinceUnixEpoch` and there should be a `Duration.ofSeconds` function.